src/pooler.c

```diff
@@ -57,6 +57,47 @@ static struct timeval err_timeout = {5, 0};
 
 static void tune_accept(int sock, bool on);
 
+static void close_socket(struct ListenSocket *ls)
+{
+	if (event_del(&ls->ev) < 0) {
+		log_warning("close_socket, event_del: %s", strerror(errno));
+	}
+	if (ls->fd > 0) {
+		safe_close(ls->fd);
+		ls->fd = 0;
+	}
+	if (pga_is_unix(&ls->addr) && cf_unix_socket_dir[0] != '@') {
+		char buf[sizeof(struct sockaddr_un) + 20];
+		snprintf(buf, sizeof(buf), "%s/.s.PGSQL.%d", cf_unix_socket_dir, cf_listen_port);
+		unlink(buf);
+	}
+	statlist_remove(&sock_list, &ls->node);
+	free(ls);
+}
+
+/*
+ * cleanup_tcp_sockets stops accepting new connections on the configured TCP
+ * sockets. This is used to support zero-downtime restarts with with
+ * so_reuseport=1. By not accepting new connections it's possible to
+ * effectively drain this process of all its connections by making sure the
+ * clients all reconnect (at their leisure). Then once they've all reconnected
+ * to another PgBouncer process, we can stop this one.
+ */
+void cleanup_tcp_sockets(void)
```

Comment:
Why only TCP sockets? I would expect all socket types to be treated equally here.

Also, do we need to do anything with libevent here, similar to cleanup_sockets? I guess if we close the socket, then no event will fire anyway, but this looks a bit odd, because it could potentially create different states where sometimes sockets are closed but still registered in libevent.

Maybe this function should just be cleanup_sockets?